packages/common/client-utils/src/layerCompat.ts

```diff
@@ -0,0 +1,144 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Result of a layer compatibility check - whether a layer is compatible with another layer.
+ * @internal
+ */
+export type LayerCompatCheckResult =
+	| { readonly isCompatible: true }
+	| {
+			readonly isCompatible: false;
+			/**
+			 * Whether the generation of the layer is compatible with the other layer.
+			 */
+			readonly isGenerationCompatible: boolean;
+			/**
+			 * The features that are required by the layer but are not supported by the other layer. This will only
+			 * be set if there are unsupported features.
+			 */
+			readonly unsupportedFeatures: string[] | undefined;
+	  };
+
+/**
+ * @internal
+ */
+export const ICompatibilityDetails: keyof IProvideCompatibilityDetails =
+	"ICompatibilityDetails";
+
+/**
+ * @internal
+ */
+export interface IProvideCompatibilityDetails {
+	readonly ICompatibilityDetails: ICompatibilityDetails;
+}
+
+/**
+ * This interface is used to communicate the compatibility details of a layer to another layer.
+ * @internal
+ */
+export interface ICompatibilityDetails extends IProvideCompatibilityDetails {
+	/**
+	 * A list of features supported by the layer at a particular layer boundary. This is used to check if these
+	 * set of features satisfy the requirements of another layer.
+	 */
+	supportedFeatures: Set<string>;
```

Comment:
can all these properties be made readonly? is there a ReadonlySet as well?

Reply:
changed them to readonly.